Refactor/dutch x

Velenir opened the pull request:
WIP

more abstraction and modularity for:
+ web3
+ Tokens
+ DutchX
+ an overarching contracts/account api in `api/index.js`

Velenir committed:
separated contracts import into a module

Velenir committed:
added preliminary types for future structures

Velenir committed:
moved windowLoaded promise to helpers and added promisify helper

Velenir committed:
added web3 wrapper instance

Velenir committed:
added Tokens abstraction

Velenir committed:
added DutchExchange abstraction

Velenir committed:
relaxed tslint rules

Velenir committed:
Default auctionIndex to last auction in getClosingPrice

Velenir committed:
changed amount argument type to Balance

Velenir committed:
added getAddress function to DutchX

Velenir committed:
changed local TokenPair type to be a limited case of global TokenPair

so that it is possible to pass {sell, buy} objects without sellAmount property

Velenir committed:
added cumulative dxAPI interface

Velenir committed:
rewrote api functions using the new promised* objects

W3stside commented on src/api/index.ts:
```diff
@@ -0,0 +1,103 @@
+}
+
+// TODO: remove ['ETH', 'GNO'] default, use actions for this
+export const getTokenBalances = async (codeList: TokenCode[] = ['ETH', 'GNO'], account?: Account) => {
```
why is this called codeList and not tokenList?

Velenir replied:
don't know, will change to `tokenList`

W3stside commented on src/api/DutchX.ts:
```diff
@@ -0,0 +1,142 @@
+
+    const [num, den] = getExchange(pair).closingPrices(index)
+
+    return num.div(den).toString()
```
Not super positive with the changes that will come - but I think we should be careful of this return of `num/den` as we may need to divide the inverse when calculating the other side of Token Exchanges _e.g `ETH/GNO` && `GNO/ETH`_

W3stside commented on src/api/DutchX.ts:
```diff
@@ -0,0 +1,142 @@
+
+    const balance = await getExchange(pair).sellerBalances(index, account)
+
+    return await balance.toString()
```
why `await` here?

Velenir replied:
nice spotting skills. will remove

W3stside commented on src/api/web3Provider.ts:
```diff
@@ -0,0 +1,63 @@
+    getETHBalance,
+    getNetwork,
+    isConnected,
+    get currentProvider() {
```
This is new to me - are you declaring the function here _and_ calling the getter?

Velenir replied:
declaring the getter, same as
```
class CL {
  get prop () {
    return this.prop
  }
}
```

Velenir committed:
renamed codeList to tokenList

Velenir committed:
removed unnecessary await

Velenir committed:
Removed DutchExchangeFactory require as it is no longer compiled

Velenir committed:
added explicit types to contractMaps

Velenir committed:
fixed wrond import

Velenir committed:
moved setProvider  into contracts.init function

can't deploy before setting providers, though can swith providers after deploying

Velenir committed:
changed tests to work by getting contracts through the escape hatch (api/contracts)

Velenir committed:
added default gas to transcations that rund out of gas otherwise

Velenir committed:
added getTokeBalance function forgetting specific token's balance

Velenir committed:
changed "seller can submit order" test to use the new api

Velenir committed:
added typings for bignumber.js

Velenir committed:
changed api functions to return BigNumbers for later conversion on demand

Velenir committed:
use closingPrice api function instead of directly calling contracts

Velenir committed:
use getTokenBalance api function instead of directly calling contracts

Velenir committed:
use getTotalBalance api/tokens function instead of directly calling contracts

Velenir committed:
removed unnecessary token transfer approval that is now taken care of inside api

Velenir committed:
use approve and transfer api/tokens function instead of directly calling contracts

Velenir committed:
added getAuctionStart function to api/DutchX

Velenir committed:
use getAuctionIndex, getSellVolumeCurrent, getSellVolumeNext and getAuctionStart api/DutchX functions instead of calling contracts directly

Velenir committed:
fixed wrong return type

Velenir committed:
fixed wrong method call

Velenir committed:
use getPrice, getBuyVolume, getSellerBalances, getBuyerBalances and getClaimedAmounts api/DutchX functions instead of direct contract calls

Velenir committed:
chnged some function names from plural to singular

Velenir committed:
use claimSellerFunds, claimBuyerFunds and postBuyOrder api/DutchX functions instead of calling contracts directly

Velenir committed:
implemented workaround for uint casting / rounding of claimed token

Velenir committed:
replaced old api/dutchx with the new more abstracted and modular API

Velenir committed:
made TokenPair.sellAmount prop optional

Velenir committed:
changed actions to use the new contracts api

Velenir committed:
changed BuyButton to use the new contracts api

Velenir committed:
changed currently unused components to the new api but mostly to fix TS errors

Velenir committed:
removed test that relied on the old api

W3stside commented on src/api/contracts.ts:
```diff
@@ -0,0 +1,38 @@
+  'TokenGNO',
+]
+
+const Contracts = contractNames.map(name => TruffleContract(require(`../../build/contracts/${name}.json`)))
```
Could you rename to `ContractJSON` or `ContractArtifacts` ? 😅

Velenir replied:
These are not contract artifacts, they are contract classes, if you will. Since each is `TruffleContract(artifact)`

W3stside replied:
oh I didn't see the TruffleContract before, sneaky. Nvm then

W3stside commented on src/api/utils/index.ts:
```diff
@@ -0,0 +1,20 @@
+})
+
+
+export const promisify = (func: Function, context: object, ...defArgs: any[]) =>
```
I hate this function 😂 😂 😂 😂

Velenir replied:
I'm open to any changes. This one is actually a slightly modified `util.promisify` from node, but we can't use node system libraries in the browser(

W3stside replied:
its fine for now

W3stside commented on src/components/BalanceButton/index.tsx:
```diff
@@ -1,6 +1,6 @@
 /* eslint no-console:0 */
```
can you delete this component, not used

Velenir committed:
removed no longer used component

W3stside reviewed (approved):
From testing and going through code it looks good, merge
